pkg/gameserverallocations/controller.go

```diff
@@ -328,8 +342,95 @@ func (c *Controller) applyMultiClusterAllocation(gsa *v1alpha1.GameServerAllocat
 // an http call to allocation service in that cluster.
 func (c *Controller) allocateFromRemoteCluster(gsa v1alpha1.GameServerAllocation, connectionInfo *multiclusterv1alpha1.ClusterConnectionInfo, namespace string) (*v1alpha1.GameServerAllocation, error) {
-	// TODO: implement getting secrets and making rest call to remote cluster
-	return nil, nil
+	var gsaResult v1alpha1.GameServerAllocation
+
+	// TODO: handle converting error to apiserver error
+	client, err := c.createRemoteClusterRestClient(namespace, connectionInfo.SecretName)
+	if err != nil {
+		return nil, err
+	}
+
+	// Forward the game server allocation request to another cluster,
+	// and disable multicluster settings to avoid the target cluster
+	// forward the allocation request again.
+	gsa.Spec.MultiClusterSetting.Enabled = false
+	body, err := json.Marshal(gsa)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: Retry on transient error --> response.StatusCode >= 500
+	response, err := client.Post(connectionInfo.AllocationEndpoint, "application/json", bytes.NewBuffer(body))
+	if err != nil {
+		return nil, err
+	}
+	defer response.Body.Close() // nolint: errcheck
+
+	data, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		return nil, err
+	}
+	if response.StatusCode >= 400 {
+		return nil, errors.New(string(data))
+	}
+
+	err = json.Unmarshal(data, &gsaResult)
+	if err != nil {
+		return nil, err
+	}
+
+	return &gsaResult, nil
+}
+
+// createRemoteClusterRestClient creates a rest client with proper certs to make a remote call.
+func (c *Controller) createRemoteClusterRestClient(namespace, secretName string) (*http.Client, error) {
+	clientCert, clientKey, caCert, err := c.getClientCertificates(namespace, secretName)
+	if err != nil {
+		return nil, err
+	}
+
+	// Load client cert
+	cert, err := tls.X509KeyPair(clientCert, clientKey)
+	if err != nil {
+		return nil, err
+	}
+
+	tlsConfig := &tls.Config{Certificates: []tls.Certificate{cert}}
+	if len(caCert) != 0 {
+		// Load CA cert, if provided and trust the server certificate.
+		// This is required for self-signed certs.
+		tlsConfig.RootCAs = x509.NewCertPool()
+		ca, err := x509.ParseCertificate(caCert)
+		if err != nil {
+			return nil, err
+		}
+		tlsConfig.RootCAs.AddCert(ca)
+	}
+
+	// Setup HTTPS client
+	return &http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: tlsConfig,
+		},
+	}, nil
+}
+
+// getClientCertificates returns the client certificates and CA cert for remote allocation cluster call
+func (c *Controller) getClientCertificates(namespace, secretName string) (clientCert, clientKey, caCert []byte, err error) {
+	secret, err := c.secretGetter.Secrets(namespace).Get(secretName, metav1.GetOptions{})
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	if secret == nil || len(secret.Data) == 0 {
+		return nil, nil, nil, fmt.Errorf("secert %s does not have data", secretName)
+	}
+
+	// Create http client using cert
+	clientCert = secret.Data[secretClientCertName]
```

Comment:
perhaps:

```
if clientCert == nil {
  return .. fmt.Erroff("missing client certificate in secret ...")
}
```

for all 3 fields

Reply:
X509KeyPair returns an error if any of the key pair is not conforming with the accepted format including null. I can add an extra error message for clarification.